### ISSUE TITLE
Ability to flag exports as tree-shakable

### ISSUE THREAD

**Issue by isidrok** [rollup/rollup#3701]
### Feature Use Case

I'm using Rollup to bundle a web app in several steps. One of them consist in bundling all third party libraries used by the application in order to cache the result and speed up the application build process.

Since libraries are bundled independently tree-shake cannot be applied and all exports are included.  

### Feature Proposal

It would be nice to have a plugin hook that is invoked for each imported module and allows flagging some of its exports as tree-shakable.

**Comment by lukastaegert** [comment]
Wouldn't `moduleSideEffects` solve this? It is available both as a user-facing option https://rollupjs.org/guide/en/#treeshake as was well as a flag when resolving/loading/transforming modules https://rollupjs.org/guide/en/#transform

**Comment by lukastaegert** [comment]
Or is this about saying: When bundling, do not include all exports of my entry point but only `foo` and `bar` and ignore `baz`?

**Comment by isidrok** [comment]
> Or is this about saying: When bundling, do not include all exports of my entry point but only `foo` and `bar` and ignore `baz`?

That's it. If we have the following code:

```js
// lib.js
export const a = 1;
export const b = 2;
```

I would like to tell Rollup that export `b` is unused and can be tree-shaken.

**Comment by lukastaegert** [comment]
Achieving that is easy by replacing the input with a file that just reexports the exports you want. A plugin could tweak the options to replace `input` with a virtual module that does just that. The bigger problem will be getting that information, maybe we can add it to the bundle object.

**Comment by lukastaegert** [comment]
Until then, using `this.parse` in the `generateBundle` hook to manually get the imports is always an option.

**Comment by isidrok** [comment]
Yeah I think that's how [this plugin]( https://github.com/pikapkg/snowpack/blob/master/packages/snowpack/src/rollup-plugins/rollup-plugin-wrap-install-targets.ts) works but thought there would be some internal information Rollup could expose to make things easier. Maybe this is an enhancement request rather than a feature proposal.

Regarding the use of `this.parse` I was going to manually parse source files in a separate step so thanks for the hint!
